src/tir/analysis/device_constraint_utils.cc

```diff
@@ -0,0 +1,523 @@
+    Buffer new_buffer = Subst(new_buffer_store->buffer.get());
+    if (!new_buffer.same_as(new_buffer_store->buffer)) {
+      return BufferStore(new_buffer, new_buffer_store->value, new_buffer_store->indices,
+                         new_buffer_store->span);
+    }
+    return new_buffer_store;
+  }
+
+  Stmt VisitStmt_(const BufferRealizeNode* buffer_realize_node) final {
+    BufferRealize new_buffer_realize =
+        Downcast<BufferRealize>(StmtExprMutator::VisitStmt_(buffer_realize_node));
+    Buffer new_buffer = Subst(new_buffer_realize->buffer.get());
+    if (!new_buffer.same_as(new_buffer_realize->buffer)) {
+      return BufferRealize(new_buffer, new_buffer_realize->bounds, new_buffer_realize->condition,
+                           new_buffer_realize->body, new_buffer_realize->span);
+    }
+    return new_buffer_realize;
+  }
+
+  // IfThenElseNode default ok
+  // AssertStmtNode default ok
+  // ProducerStoreNode default ok (though does not visit producer)
+  // ProducerRealizeNode default ok (though does not visit producer)
+
+  Stmt VisitStmt_(const PrefetchNode* prefetch_node) final {
+    Prefetch new_prefetch = Downcast<Prefetch>(StmtExprMutator::VisitStmt_(prefetch_node));
+    Buffer new_buffer = Subst(new_prefetch->buffer.get());
+    if (!new_buffer.same_as(new_prefetch->buffer)) {
+      return Prefetch(new_buffer, prefetch_node->bounds, prefetch_node->span);
+    }
+    return new_prefetch;
+  }
+
+  // SeqStmtNode default ok
+  // EvaluateNode default ok
+
+  BufferRegion VisitItem(const BufferRegionNode* buffer_region_node) {
+    Buffer new_buffer = Subst(buffer_region_node->buffer.get());
+    if (!new_buffer.same_as(buffer_region_node->buffer)) {
+      return BufferRegion(new_buffer, buffer_region_node->region);
+    }
+    return GetRef<BufferRegion>(buffer_region_node);
+  }
+
+  MatchBufferRegion VisitItem(const MatchBufferRegionNode* match_buffer_region_node) {
+    // The source field has a referential occurrence of the  buffer. Apply the buffer substitution
+    // to that.
+    BufferRegion new_source = VisitItem(match_buffer_region_node->source.get());
+    // The buffer field however is a definitional occurrence, aliased on top of the source.
+    // Transfer any memory scope from the source to the destination.
+    Optional<SEScope> opt_se_scope = GetBufferConstraint(new_source->buffer);
+    tir::Buffer new_buffer;
+    if (opt_se_scope.defined()) {
+      new_buffer = RewriteBuffer(match_buffer_region_node->buffer, opt_se_scope.value());
+    } else {
+      new_buffer = match_buffer_region_node->buffer;
+    }
+    if (!new_buffer.same_as(match_buffer_region_node->buffer) ||
+        !new_source.same_as(match_buffer_region_node->source)) {
+      return MatchBufferRegion(new_buffer, new_source);
+    }
+    return GetRef<MatchBufferRegion>(match_buffer_region_node);
+  }
+
+  template <typename T>
+  Array<T> VisitItems(Array<T> items) {
```

Comment:
What's the diff between this method and Array's MutateByApply API?

Reply:
ah, look at that. switched, thx.